File: backend/user-service/src/main/java/com/bookmyshow/user/service/AuthService.java
package com.bookmyshow.user.service;

import com.bookmyshow.common.exception.InvalidTokenException;
import com.bookmyshow.user.dto.LoginRequest;
import com.bookmyshow.user.dto.LoginResponse;
import com.bookmyshow.user.dto.UserProfileResponse;
import com.bookmyshow.user.entity.RefreshToken;
import com.bookmyshow.user.entity.User;
import com.bookmyshow.user.repository.RefreshTokenRepository;
import com.bookmyshow.user.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Service
@Transactional
@RequiredArgsConstructor
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final JwtTokenProvider jwtTokenProvider;
    private final UserService userService;
    private final RefreshTokenRepository refreshTokenRepository;

    public LoginResponse login(LoginRequest request) {
        authenticationManager.authenticate(
            new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        User user = userService.getUserByEmail(request.getEmail());

        String accessToken = jwtTokenProvider.generateAccessToken(user.getEmail(), user.getId());
        String refreshToken = jwtTokenProvider.generateRefreshToken();

        saveRefreshToken(user.getId(), refreshToken);
        userService.updateLastLogin(user.getId());

        UserProfileResponse userProfile = new UserProfileResponse(
            user.getId(),
            user.getEmail(),
            user.getFirstName(),
            user.getLastName(),
            user.getPhoneNumber(),
            user.getRole(),
            user.getIsActive(),
            user.getIsEmailVerified(),
            user.getCreatedAt(),
            user.getLastLoginAt()
        );

        return new LoginResponse(
            accessToken,
            refreshToken,
            jwtTokenProvider.getAccessTokenExpiration(),
            userProfile
        );
    }

    public LoginResponse refreshAccessToken(String refreshTokenString) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenString)
            .orElseThrow(() -> new InvalidTokenException("Invalid refresh token"));

        if (refreshToken.getExpiresAt().isBefore(LocalDateTime.now())) {
            refreshTokenRepository.delete(refreshToken);
            throw new InvalidTokenException("Refresh token has expired");
        }

        User user = userService.getUserById(refreshToken.getUserId());

        String newAccessToken = jwtTokenProvider.generateAccessToken(user.getEmail(), user.getId());

        UserProfileResponse userProfile = new UserProfileResponse(
            user.getId(),
            user.getEmail(),
            user.getFirstName(),
            user.getLastName(),
            user.getPhoneNumber(),
            user.getRole(),
            user.getIsActive(),
            user.getIsEmailVerified(),
            user.getCreatedAt(),
            user.getLastLoginAt()
        );

        return new LoginResponse(
            newAccessToken,
            refreshTokenString,
            jwtTokenProvider.getAccessTokenExpiration(),
            userProfile
        );
    }

    public void logout(String refreshToken) {
        refreshTokenRepository.deleteByToken(refreshToken);
    }

    private void saveRefreshToken(Long userId, String token) {
        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setUserId(userId);
        refreshToken.setToken(token);
        refreshToken.setExpiresAt(
            LocalDateTime.now().plusSeconds(jwtTokenProvider.getRefreshTokenExpiration() / 1000)
        );
        refreshTokenRepository.save(refreshToken);
    }
}

File: backend/user-service/app/services/auth_service.py
"""Authentication service for business logic."""

from datetime import datetime, timedelta

from bookmyshow_common.exceptions import InvalidTokenException, UnauthorizedException
from bookmyshow_common.security import JwtTokenProvider, PasswordHasher

from app.models import RefreshToken
from app.repositories.user_repository import UserRepository
from app.schemas.auth import LoginRequest, LoginResponse
from app.schemas.user import UserProfileResponse
from app.services.user_service import UserService


class AuthService:
    """Service for authentication-related business logic."""

    def __init__(
        self,
        repository: UserRepository,
        user_service: UserService,
        jwt_provider: JwtTokenProvider,
    ):
        """Initialize auth service.
        
        Args:
            repository: User repository instance
            user_service: User service instance
            jwt_provider: JWT token provider instance
        """
        self.repository = repository
        self.user_service = user_service
        self.jwt_provider = jwt_provider
        self.password_hasher = PasswordHasher()

    async def login(self, request: LoginRequest) -> LoginResponse:
        """Authenticate user and generate tokens.
        
        Args:
            request: Login request with email and password
            
        Returns:
            Login response with tokens and user profile
            
        Raises:
            UnauthorizedException: If authentication fails
        """
        # Get user by email
        user = await self.user_service.get_user_by_email(request.email)

        # Verify password
        if not self.password_hasher.verify_password(request.password, user.password_hash):
            raise UnauthorizedException("Invalid email or password")

        # Generate tokens
        access_token = self.jwt_provider.generate_access_token(user.email, user.id)
        refresh_token_str = self.jwt_provider.generate_refresh_token()

        # Save refresh token
        await self._save_refresh_token(user.id, refresh_token_str)

        # Update last login
        await self.user_service.update_last_login(user.id)

        # Refresh user to get updated last_login_at
        user = await self.user_service.get_user_by_id(user.id)

        user_profile = UserProfileResponse(
            id=user.id,
            email=user.email,
            first_name=user.first_name,
            last_name=user.last_name,
            phone_number=user.phone_number,
            role=user.role,
            is_active=user.is_active,
            is_email_verified=user.is_email_verified,
            created_at=user.created_at,
            last_login_at=user.last_login_at,
        )

        return LoginResponse(
            access_token=access_token,
            refresh_token=refresh_token_str,
            token_type="Bearer",
            expires_in=self.jwt_provider.get_access_token_expiration(),
            user=user_profile,
        )

    async def refresh_access_token(self, refresh_token_str: str) -> LoginResponse:
        """Refresh access token using refresh token.
        
        Args:
            refresh_token_str: Refresh token string
            
        Returns:
            Login response with new access token
            
        Raises:
            InvalidTokenException: If refresh token is invalid or expired
        """
        # Get refresh token from database
        refresh_token = await self.repository.get_refresh_token_by_token(refresh_token_str)
        if not refresh_token:
            raise InvalidTokenException("Invalid refresh token")

        # Check if token is expired
        if refresh_token.expires_at < datetime.utcnow():
            await self.repository.delete_refresh_token(refresh_token)
            raise InvalidTokenException("Refresh token has expired")

        # Get user
        user = await self.user_service.get_user_by_id(refresh_token.user_id)

        # Generate new access token
        new_access_token = self.jwt_provider.generate_access_token(user.email, user.id)

        user_profile = UserProfileResponse(
            id=user.id,
            email=user.email,
            first_name=user.first_name,
            last_name=user.last_name,
            phone_number=user.phone_number,
            role=user.role,
            is_active=user.is_active,
            is_email_verified=user.is_email_verified,
            created_at=user.created_at,
            last_login_at=user.last_login_at,
        )

        return LoginResponse(
            access_token=new_access_token,
            refresh_token=refresh_token_str,
            token_type="Bearer",
            expires_in=self.jwt_provider.get_access_token_expiration(),
            user=user_profile,
        )

    async def logout(self, refresh_token: str) -> None:
        """Logout user by invalidating refresh token.
        
        Args:
            refresh_token: Refresh token to invalidate
        """
        await self.repository.delete_refresh_token_by_token(refresh_token)

    async def _save_refresh_token(self, user_id: int, token: str) -> None:
        """Save refresh token to database.
        
        Args:
            user_id: User ID
            token: Refresh token string
        """
        expires_at = datetime.utcnow() + timedelta(
            milliseconds=self.jwt_provider.get_refresh_token_expiration()
        )

        refresh_token = RefreshToken(
            user_id=user_id,
            token=token,
            expires_at=expires_at,
        )

        await self.repository.save_refresh_token(refresh_token)
